api/storage_service.cc

```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
         }));
 }
 
+static
+future<json::json_return_type>
+rest_set_client_routes(http_context& ctx, sharded<service::storage_service>& ss, service::raft_group0_client& group0_client, std::unique_ptr<http::request> req) {
+    if (!ss.local().get_feature_service().client_routes) {
+        apilog.warn("set_client_routes: called before the cluster feature was enabled");
+        throw std::runtime_error("set_client_routes requires all nodes to support the CLIENT_ROUTES cluster feature");
+    }
+
+    rapidjson::Document root;
+    auto content = co_await util::read_entire_stream_contiguous(*req->content_stream);
+    root.Parse(content.c_str());
+    if (!root.IsArray()) {
+        throw bad_param_exception("Body must be a JSON array");
+    }
+
+    co_await ss.invoke_on(0, [&group0_client, &root] (service::storage_service& ss) -> future<> {
+        int retries = 10;
+        while (true) {
+            try {
+                auto batch = service::group0_batch(co_await group0_client.start_operation(ss.get_abort_source()));
+
+                for (const auto& v : root.GetArray()) {
+                    if (!v.IsObject()) { throw bad_param_exception("Each element must be object"); }
+
+                    // Required string fields
+                    for (const auto& name : {"connection_id", "host_id", "address"}) {
+                        if (!v.HasMember(name)) {
+                            throw bad_param_exception(fmt::format("Missing '{}'", name));
+                        }
+                        if (!v[name].IsString()) {
+                            throw bad_param_exception(fmt::format("'{}' must be a string", name));
+                        }
+                    }
+
+                    // Optional integer port fields; capture only those provided.
+                    std::map<sstring, std::optional<int32_t>> port_map;
```

Comment:
I'm not a fan of scattering client_routes over existing "bags of everything". Can't we create a dedicated class/service for them? clients_router/manager or something. It can take an explicit constructor dependency on `raft_group0_client` in `main.cc`. We can move system_keyspace CQL-s into it to not clutter system_keyspace.hh/cc even more than they are today.